nrf-hal-common/src/usbd/mod.rs

```diff
@@ -0,0 +1,714 @@
+//! A `usb-device` implementation using the USBD peripheral.
+//!
+//! Difficulties:
+//! * Control EP 0 is special:
+//!   * Setup stage is put in registers, not RAM.
+//!   * Different events are used to initiate transfers.
+//!   * No notification when the status stage is ACK'd.
+
+mod errata;
+
+use crate::{
+    clocks::{Clocks, ExternalOscillator},
+    pac::USBD,
+};
+use core::sync::atomic::{compiler_fence, Ordering};
+use core::cell::Cell;
+use core::mem::MaybeUninit;
+use cortex_m::interrupt::{self, Mutex};
+use usb_device::{
+    bus::{PollResult, UsbBus, UsbBusAllocator},
+    endpoint::{EndpointAddress, EndpointType},
+    UsbDirection, UsbError,
+};
+
+fn dma_start() {
+    compiler_fence(Ordering::Release);
+}
+
+fn dma_end() {
+    compiler_fence(Ordering::Acquire);
+}
+
+struct Buffers {
+    // Buffers can be up to 64 Bytes since this is a Full-Speed implementation.
+    in_lens: [u8; 9],
+    out_lens: [u8; 9],
+}
+
+impl Buffers {
+    fn new() -> Self {
+        Self {
+            in_lens: [0; 9],
+            out_lens: [0; 9],
+        }
+    }
+}
+
+unsafe impl Sync for Buffers {}
+
+#[derive(Copy, Clone)]
+enum TransferState {
+    NoTransfer,
+    Started(u16),
+}
+
+#[derive(Copy, Clone)]
+struct EP0State {
+    direction: UsbDirection,
+    remaining_size: u16,
+    in_transfer_state: TransferState,
+    is_set_address: bool,
+}
+
+/// USB device implementation.
+pub struct Usbd<'c> {
+    periph: Mutex<USBD>,
+    // argument passed to `UsbDeviceBuilder.max_packet_size_0`
+    max_packet_size_0: u16,
+    bufs: Buffers,
+    used_in: u8,
+    used_out: u8,
+    iso_in_used: bool,
+    iso_out_used: bool,
+    ep0_state: Mutex<Cell<EP0State>>,
+    busy_in_endpoints: Mutex<Cell<u16>>,
+
+    // used to freeze `Clocks` and ensure they remain in the `ExternalOscillator` state
+    _clocks: &'c (),
+}
+
+impl<'c> Usbd<'c> {
+    /// Creates a new USB bus, taking ownership of the raw peripheral.
+    ///
+    /// # Parameters
+    ///
+    /// * `periph`: The raw USBD peripheral.
+    #[inline]
+    pub fn new<L, LSTAT>(
+        periph: USBD,
+        _clocks: &'c Clocks<ExternalOscillator, L, LSTAT>,
```

Comment:
I'm testing out this branch in conjunction with [nrf-softdevice](https://github.com/embassy-rs/nrf-softdevice) with the goal of writing some BLE dongle firmware. The SoftDevice takes ownership of the clock peripheral and makes it impossible to use `nrf52840_hal::Clocks`. I use the [sd_clock_hfclk_request](https://github.com/embassy-rs/nrf-softdevice/blob/77db337aef97c58d0824b160778b100a31a46ebf/softdevice/s140/headers/nrf_soc.h#L698) function to request that the external oscillator be used and poll until it is done. Am I right in thinking the `_clock` is used here as a compile time check to ensure that this is done? Could there be another function like `new_with_clock_set` that has a runtime check to increase flexibility?